File: merge sort.cpp
#include<iostream>
#include<vector>
using namespace std;

void merge(vector<int> &arr,int s,int e){
    
    //as the array will be in two parts 
    //1st array start to mid
    //2nd array mid+1 to end
    int i=s, m=(s+e)/2, j=m+1;
    vector<int> temp;
    
    while(i<=m and j<=e){
        if (arr[i] < arr[j]){
            temp.push_back(arr[i]);
            i++;
        }
        else{
            temp.push_back(arr[j]);
            j++;
        }
    }
    
    //if first array has any remaining elements add them to temp
    while(i<=m){
        temp.push_back(arr[i]);
        i++;
    }
    
    //if second array has any remaining elements add them to temp
    while(j<=e){
        temp.push_back(arr[j]);
        j++;
    }
    
    //copy back the elements from temp to original arr
    int k=0;
    for(int i=s;i<=e;i++){
        arr[i]=temp[k++];
    }
    return;
}

void mergesort(vector<int> &arr,int s,int e){
    
    // base case
    if(s>=e){
        return;
    }
    
    //recursive case
    int mid=(s+e)/2;
    mergesort(arr,s,mid);
    mergesort(arr,mid+1,e);
    return merge(arr,s,e);
    
}

int main()
{
    vector<int> a{10,5,2,0,7,6,4};
    int s=0;
    int e=a.size()-1;
    mergesort(a,s,e);
    for(int i:a){
        cout<<i<<" ";
    }
    return 0;
}

File: Arrays/MergeSort.java
public class MergeSort {
    public static void divide(int[] nums,int low,int high){
        if(low>=high) return;
        int mid = low+(high-low)/2;
        divide(nums, low, mid);
        divide(nums, mid+1, high);
        merge(nums, low, mid, high);
    }

    public static void merge(int[] nums, int low, int mid, int high){
        int i=low, j=mid+1, k=0;
        int[] temp = new int[high - low + 1];
        while(i<=mid && j<=high){
            if(nums[i]<=nums[j]){
                temp[k]=nums[i];
                k++;
                i++;
            }
            else{
                temp[k]=nums[j];
                k++;
                j++;
            }
        }
        while(i<=mid){
            temp[k]=nums[i];
            k++;
            i++;
        }
        while(j<=high){
            temp[k]=nums[j];
            k++;
            j++;
        }
        for (int p = 0; p < temp.length; p++) {
            nums[low + p] = temp[p];
        }
    }

    public static void main(String args[]) {
        int n = 7;
        int arr[] = { 9, 4, 7, 6, 3, 1, 5 };
        System.out.println("Before sorting array: ");
        for (int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
        divide(arr, 0, n - 1);
        System.out.println("After sorting array: ");
        for (int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}
